### PR TITLE
Ruby implementation of finding anagrams

### PR DESCRIPTION
**Pull request by zzokki81**

### CODE REVIEW

**Review by basti — CHANGES_REQUESTED** [review]
@zzokki81 ok, there's too much Go influence here. Look at the Anagrams class. `print_anagrams` is the only really callable method, so it's the only one that should be public. It calls `group_anagrams` that return certain processed data, which it got from`read_file` that returns file contents. Each is just using return value - typical functions: `print_anagrams(group_anagrams(read_file)))`.

Try to modify the class so it actually behaves like an object. While you're at it add a method to find anagrams for one given word. This will help you realize that you need to have some object state.

**Comment by basti** [comment]
@zzokki81 apologies, I should have looked at this sooner. Can you remove .bundle from git repo? Put it in .gitignore. Currently, it says you changed 288 files.

**Comment by basti** [comment]
@zzokki81 you didn't actually remove it: https://github.com/zzokki81/anagrams/pull/1/files - when you add something to .gitignore if that is already committed it won't be automatically removed. You need to git rm it.

**Comment by zzokki81** [comment]
> @zzokki81 you didn't actually remove it: https://github.com/zzokki81/anagrams/pull/1/files - when you add something to .gitignore if that is already committed it won't be automatically removed. You need to git rm it.

Ok, my mistake. I think it's ok now.

File: findanagram.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+require_relative 'lib/anagrams'
+
+input_word = ARGV[0]
+text_file = ARGV[1]
```
Comment (basti) [inline findanagram.rb]: @zzokki81 can you please modify the program so I can give it a dictionary filename and enter any number of the words I want anagrams for. For example, `ruby findanagrams.rb dict.txt tapes pictures undress`.

**Review by basti — CHANGES_REQUESTED** [review]
@zzokki81 still asking for changes that will hopefully push you toward a more objected design without me explicitly telling you what to do.

**Review by basti — CHANGES_REQUESTED** [review]
@zzokki81 ok, let's try to be more explicit. There should be only one class defined. An instance of the class should analyze the dictionary only once and then be able to perform all the required actions. 

Speaking of which, I want to be able to do the following:

```
# create an object by passing a filename
anagrams = Anagrams.new(dictionary_filename)

# but also create an object by passing an array
# (useful for testing and when you have words in a database or something other than a file)
anagrams = Anagrams.new(["array", "of", "words"])

#######

# return all anagrams as an array of arrays
# (e.g. [ ["ab", "ba"], ["abc", "bca", "cba"], ...]
anagrams.all

# return all anagrams as a string
anagrams.to_s

# return an array of words that are anagrams of the given word
anagrams.find(given_word)
```

If you have any questions let me know.

File: lib/anagrams.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
+class Anagrams
+  def initialize(anagram_file)
+    @anagram_file = anagram_file
+  end
+
+  def group_anagrams
+    anagram_group = {}
+    read_file.group_by do |word|
+      sorted_word = word.chars.sort.join
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 I don't think this will work for anagrams such as `ABS` and `AB's`. You should ignore case and non-letter chars. Try adding some tests and you'll probably see them fail and not match words with special chars or upper/lower case letters.

File: findanagram.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+require_relative('lib/anagrams.rb')
+
+text_file, *input_word = ARGV
+
+input_word.each do |word|
+  anagram = Anagrams.new(text_file)
```
Comment (basti) [inline findanagram.rb]: @zzokki81 I explicitly said:

> An instance of the class should analyze the dictionary only once and then be able to perform all the required actions.

Why are you creating an object each time you want to get anagrams of a word? Also, you are not analyzing the dictionary only once.

This results in slow execution:

```
$ ruby findanagram.rb anagrams-wordlist.txt undo undo undo undo             *[feature/anagrams] 
Anagrams for given word 'undo' is: 
udon undo
0.881240999995498
Anagrams for given word 'undo' is: 
udon undo
0.8340949999983422
Anagrams for given word 'undo' is: 
udon undo
0.812352999986615
Anagrams for given word 'undo' is: 
udon undo
0.8182730000116862
```

For each word it takes just shy of a second to get anagrams - even if it's the same word.

File: lib/anagrams.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  def find_anagram(input_word)
+    result = []
+    word_characters = input_word.chars.to_a.sort.join
+    read_file.each do |word|
+      result.push(word) if word.chars.to_a.sort.join == word_characters
+    end
+    result
+  end
+
+  def to_string
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 requirement was for this method to be called `to_s`.

File: lib/anagrams.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+class Anagrams
+  def initialize(anagram_file)
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 `anagram_file` is no longer a good name for this var. At the least it should be something more intention revealing like `dictionary` or `words`, you can even go overboard (it's just a local var/argument) and call it `words_array_or_file`.

**Review by basti — CHANGES_REQUESTED** [review]

File: findanagram.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+finder = Anagrams.new(dictionary)
+input_word.uniq.each do |word|
+anagram_result = finder.find_anagram(word)
+   anagram_result.delete(word)
+  if anagram_result.empty?
+    puts("No anagrams found for word '#{word}'")
+  else
+    puts("Anagrams for given word '#{word}' is: ")
+    puts((anagram_result * ' '))
+  end
```
Comment (basti) [inline findanagram.rb]: @zzokki81 fix indentation in this block.

File: findanagram.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+require_relative('lib/anagrams.rb')
+
+dictionary, *input_word = ARGV
+finder = Anagrams.new(dictionary)
+input_word.uniq.each do |word|
+anagram_result = finder.find_anagram(word)
```
Comment (basti) [inline findanagram.rb]: @zzokki81 https://github.com/zzokki81/anagrams/pull/1#pullrequestreview-460989282 - this should have been `find` method. In which case `finder.find` would expose an inadequate name for the Anagrams class object. Usually, you would name a variable containing an instance of class something indicating that. In this case, I'd go for `anagrams` or `anagram_finder` and you would have `anagrams.find` which reads nicely.

Please check all other methods, because I think you didn't follow the requirements I laid out in that comment.

File: lib/anagrams.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  def read_file
+    if @dictionary.kind_of?(Array)
+      @dictionary
+    else
+      File.read(@dictionary).encode('UTF-8', invalid: :replace, replace: '').split("\n")
+    end
+  end
+
+  def sort_and_validate(word)
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 misleading name. This method doesn't sort a word (well, at least not really as you cannot sort a word) and it definitely doesn't perform any validation. A better name would be `word_hash` because it generates a string that identifies a word (kind of), `word_key` because it generates a key that represents a word which can be used for grouping words, or something similar.

File: lib/anagrams.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+  def to_s
+    group_anagrams.each do |anagram_block|
+      puts(anagram_block.join(' ')) if anagram_block.size > 1
+    end.to_s
+  end
+
+  private
+
+  def read_file
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 misleading name. You are really returning an array of words, a dictionary. Reading a file is just one possible thing you are doing in this method.

File: lib/anagrams.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    if @dictionary.kind_of?(Array)
+      @dictionary
+    else
+      File.read(@dictionary).encode('UTF-8', invalid: :replace, replace: '').split("\n")
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 with the current setup, whenever you call `read_file` you are going to have to open file, read it, put it's content in the array. That means when you call `find_anagrams` multiple times on the same object that for each call you load words over and over. Instead, you should cache the dictionary. It simple as:
```suggestion
      @dictionary = File.read(@dictionary).encode('UTF-8', invalid: :replace, replace: '').split("\n")
```

Next time you call this method `@dictionary` will be an array and you won't load file again.

Also, you should find a way to prevent an exception when this happens:

```
$ ruby findanagram.rb anagrams-wordlist-DOESNOTEXIST.txt Abs ABS abs
Traceback (most recent call last):
	5: from findanagram.rb:5:in `<main>'
	4: from findanagram.rb:5:in `each'
	3: from findanagram.rb:6:in `block in <main>'
	2: from /Users/basti/code/lab/apprenticeship/zoran_rebic/anagrams/lib/anagrams.rb:19:in `find_anagram'
	1: from /Users/basti/code/lab/apprenticeship/zoran_rebic/anagrams/lib/anagrams.rb:37:in `read_file'
/Users/basti/code/lab/apprenticeship/zoran_rebic/anagrams/lib/anagrams.rb:37:in `read': No such file or directory @ rb_sysopen - anagrams-wordlist-bad.txt (Errno::ENOENT)
```

File: lib/anagrams.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+      (anagram_group[sorted_word] << word)
+    end
+    anagram_group.values
+  end
+
+  def find_anagram(input_word)
+    result = []
+    word_characters = sort_and_validate(input_word)
+    read_file.each do |word|
+      result.push(word) if sort_and_validate(word) == word_characters
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 whenever `find_anagram` is called you process the whole dictionary and check each word separately. That means for 1 word you will load each dictionary word once, process it, and then compare it. For the dictionary of 1000 words, you need to do this 1001 times (+1 is for a given word). For 50 words it'll be 1000 * 50 + 1 = 50001 processed words.

```
$ wc -l anagrams-wordlist.txt
  338882 anagrams-wordlist.txt
```

This means there are 338882 lines in that file: 338,882*50+1=16,944,101 processed words. I doubt that this is faster than having those words cached in a hash.

Please try to cache processed words in a hash so that subsequent `find_anagrams` calls don't have to reprocess the dictionary. You should use the same cache for other methods.

File: lib/anagrams.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    end
+    result
+  end
+
+  def to_s
+    group_anagrams.each do |anagram_block|
+      puts(anagram_block.join(' ')) if anagram_block.size > 1
```
Comment (basti) [inline lib/anagrams.rb]: @zzokki81 `to_s` means return object string representation. When called I would expect to get a string not to do print those strings to the output.

File: spec/anagrams_spec.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+describe Anagrams do
+  describe '.group_anagrams' do
+    it 'Returns anagrams from a file: ' do
+      anagram = Anagrams.new('AnagramFile.txt')
+      expected_anagrams = [%w[spate tapes pates septa peats tepas paste],
+                           %w[cuprites pictures crepitus piecrust], %w[undress sunders]]
+      expect(anagram.group_anagrams).to eql expected_anagrams
+    end
+
+    it 'Returns anagrams from array: ' do
```
Comment (basti) [inline spec/anagrams_spec.rb]: @zzokki81 this outputs quite unreadable and unusable print out:

```
$ rspec -fd

Anagrams
  .group_anagrams
    Returns anagrams from a file:
    Returns anagrams from array:
```

`describe` and `it` descriptions should resemble sentences as much as possible.

Beside that... 1) You need to add specs for other public methods in Anagrams class. 2) Whenever I report a "bug", like "ABC and abc are not detected as anagrams" you should add a failing test and then fix the code. Please add specs for these edge cases I reported earlier. That way test suite grows and covers more cases over time.

**Review by basti — CHANGES_REQUESTED** [review]
@zzokki81 moving in the right direction, but there's still a lot of work to bring this into good shape.

**Review by basti — APPROVED** [review]
